src/codegen.cpp

```diff
@@ -2842,8 +2858,10 @@ static jl_cgval_t emit_call(jl_expr_t *ex, jl_codectx_t *ctx)
                 else {
                     fval = emit_expr(args[0], ctx);
                 }
-                if (ctx->linfo->def) // root this li in case it gets deleted from the cache in `f`
-                    jl_add_linfo_root(ctx->linfo, (jl_value_t*)li);
+                // keep a reference to the called function's roots array in case
+                // the method is replaced (PR #14301)
+                if (ctx->linfo->def && li->def && li->def->roots)
+                    jl_add_linfo_root(ctx->linfo, (jl_value_t*)li->def->roots);
```

Comment:
I was thinking this would be a decent opportunity to incrementally start to add the entry world counters (min / max). an entry would be invisible if the world counter is not between min (this would be monotonically incrementing on every method add and copied from every method definition to its cache entries) and max (initialized to typemax). initially, the TLS world counter can just be equal to the method add counter.